File: 2024-01-04/bubble_sort.go
package main

import "fmt"

// とりあえず書いてみたバブルソート
func bubbleSort(param []int) []int {
	target := make([]int, len(param))
	copy(target, param)
	for i := 0; i < len(target); i++ {
		for k := i; k < len(target); k++ {
			if target[i] > target[k] {
				tmp := target[i]
				target[i] = target[k]
				target[k] = tmp
			}
		}
	}
	return target
}

// goの文法を活かして書いてみたバブルソート
func sortWithBubble(param []int) []int {
	target := make([]int, len(param))
	copy(target, param)
	n := len(target)
	for i := 0; i < n; i++ {
		for j := 0; j < n-i-1; j++ {
			if target[j] > target[j+1] {
				target[j], target[j+1] = target[j+1], target[j]
			}
		}
	}
	return target
}

func main() {
	param := []int{5, 2, 11, 7, 3, 13}
	result := bubbleSort(param)
	fmt.Println(result)
	expected := [6]int{2, 3, 5, 7, 11, 13}
	fmt.Println(expected)

	param = []int{5, 2, 11, 7, 3, 13}
	result = sortWithBubble(param)
	fmt.Println(result)
	expected = [6]int{2, 3, 5, 7, 11, 13}
	fmt.Println(expected)
}

File: 2023/2023-12-30/bubble_sort.py
from unittest import TestCase
from typing import TypeVar, Any, Protocol, Sequence
import unittest


class Comparable(Protocol):
    """比較可能を表すプロトコル"""

    def __eq__(self, __value: Any) -> bool:
        ...

    def __lt__(self, __value: Any) -> bool:
        ...

    def __le__(self, __value: Any) -> bool:
        ...

    def __gt__(self, __value: Any) -> bool:
        ...

    def __ge__(self, __value: Any) -> bool:
        ...


T = TypeVar("T", bound=Comparable)


def bubble_sort(target: Sequence[T]) -> list[T]:
    result = list(target)
    size = len(result)
    for i in range(size):
        for k in range(i + 1, size):
            if result[i] > result[k]:
                tmp = result[i]
                result[i] = result[k]
                result[k] = tmp
    return result


class TestClass(TestCase):
    def test_bubble_sort(self):
        param = [-1, 2, 10, 3, -23, 3]
        expected = [-23, -1, 2, 3, 3, 10]
        actual = bubble_sort(param)
        self.assertEqual(expected, actual)

    def test_bubble_sort_str(self):
        param = ["d", "b", "c", "a", "e"]
        expected = ["a", "b", "c", "d", "e"]
        actual = bubble_sort(param)
        self.assertEqual(expected, actual)

    def test_bubble_sort_tuple(self):
        param = (1, "a", 3)
        with self.assertRaises(TypeError):
            bubble_sort(param)


if __name__ == "__main__":
    unittest.main()
